osquery/events/windows/usn_journal_reader.cpp

```diff
@@ -0,0 +1,893 @@
+                                     const USNFileReferenceNumber& ref) {
+  file_id = {};
+  file_id.dwSize = sizeof(FILE_ID_DESCRIPTOR);
+
+  if (ref.data.size() <= sizeof(LARGE_INTEGER)) {
+    file_id.Type = FileIdType;
+    std::copy(ref.data.begin(),
+              ref.data.end(),
+              reinterpret_cast<std::uint8_t*>(&file_id.FileId.QuadPart));
+  } else {
+    /* NOTE(ww): osquery retains Windows 7 compatibility,
+       which means that we don't have access to the FILE_ID_128
+       member of FILE_ID_DESCRIPTOR's id union. We use GUID instead,
+       since it's the same width and the actual typing probably won't
+       matter for our purposes.
+     */
+    file_id.Type = ObjectIdType;
+    std::copy(ref.data.begin(),
+              ref.data.end(),
+              reinterpret_cast<std::uint8_t*>(&file_id.ObjectId));
+  }
+}
+
+namespace USNParsers {
+bool GetUpdateSequenceNumber(USN& usn, const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  switch (record->MajorVersion) {
+  case 2U:
+    usn = reinterpret_cast<const USN_RECORD_V2*>(record)->Usn;
+    return true;
+
+  case 3U:
+    usn = reinterpret_cast<const USN_RECORD_V3*>(record)->Usn;
+    return true;
+
+  default:
+    return false;
+  }
+}
+
+bool GetFileReferenceNumber(USNFileReferenceNumber& ref_number,
+                            const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  switch (record->MajorVersion) {
+  case 2U: {
+    ref_number =
+        reinterpret_cast<const USN_RECORD_V2*>(record)->FileReferenceNumber;
+    return true;
+  }
+
+  case 3U: {
+    ref_number =
+        reinterpret_cast<const USN_RECORD_V3*>(record)->FileReferenceNumber;
+    return true;
+  }
+
+  default:
+    return false;
+  }
+}
+
+bool GetParentFileReferenceNumber(USNFileReferenceNumber& ref_number,
+                                  const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  switch (record->MajorVersion) {
+  case 2U: {
+    ref_number = reinterpret_cast<const USN_RECORD_V2*>(record)
+                     ->ParentFileReferenceNumber;
+    return true;
+  }
+
+  case 3U: {
+    ref_number = reinterpret_cast<const USN_RECORD_V3*>(record)
+                     ->ParentFileReferenceNumber;
+    return true;
+  }
+
+  default:
+    return false;
+  }
+}
+
+bool GetTimeStamp(std::time_t& record_timestamp, const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  LARGE_INTEGER update_timestamp = {};
+
+  switch (record->MajorVersion) {
+  case 2U:
+    update_timestamp =
+        reinterpret_cast<const USN_RECORD_V2*>(record)->TimeStamp;
+    break;
+
+  case 3U:
+    update_timestamp =
+        reinterpret_cast<const USN_RECORD_V3*>(record)->TimeStamp;
+    break;
+
+  default:
+    return false;
+  }
+
+  // Convert the timestamp to seconds and rebase it (from 1601 to the epoch
+  // time)
+  record_timestamp =
+      static_cast<std::time_t>(update_timestamp.QuadPart / 10000000ULL) -
```

Comment:
@muffins this is an example of the "windows time constant" I talked about a while back. Are there utility functions for this conversion?

Reply:
I don't see a helper function in the Windows time utilities; I could add one if desired 🙂

Reply:
That's ok, we can follow up and add it.